### PR TITLE
Optional registry

### PR DESCRIPTION
**Pull request by tdamsma**
Implements changes suggested in #485. 

I already put it up here for review, but did not properly document the changes yet.

### CODE REVIEW

**Comment by stevepiercy** [comment]
This is looking good! Just a couple more lines to [cover](https://travis-ci.org/github/Pylons/deform/jobs/740794017#L344). Lint and docs don't like the docstrings, so I'll suggest improvements to fix those jobs.

File: deform/widget.py
Comment (stevepiercy) [inline deform/widget.py:145]:
```suggestion
        The requirements are specified as a sequence of either:
```

File: deform/widget.py
Comment (stevepiercy) [inline deform/widget.py:150]: The above needs to be formatted for reST and Black for linting. Leading and trailing blank lines are needed for reST lists. For Black line wraps must be less than 79 characters in docstrings. In narrative text, we don't use line wraps.

```

            1.  two-tuples in the form ``(requirement_name, version_id)``.
                Using the resource registry the **logical** requirement name
                identifiers are resolved to concrete files using the
                ``resource_registry``, or
            2.  dicts that concretely point to resources such as ``{"js":
                "deform:static/tinymce/tinymce.min.js"}``.

```

**Review by stevepiercy — CHANGES_REQUESTED** [review]
These changes should get lint and docs to pass.

File: deform/widget.py
Comment (tdamsma) [reply deform/widget.py]: thanks, I never write reST

**Comment by stevepiercy** [comment]
@tdamsma the functional tests started failing after the merge commit https://github.com/Pylons/deform/pull/486/commits/51b5418c4a78b035f1b95711cf82edbb2bb50dbb. Can you take a look to see what needs fixing?

**Comment by tdamsma** [comment]
yes, I'll have a look. I switched computer three times in the last week, so need to take a moment to get a proper environment up and running. I can't see why/how [51b5418](https://github.com/Pylons/deform/commit/51b5418c4a78b035f1b95711cf82edbb2bb50dbb) could have changed anything for the functional tests though

**Comment by stevepiercy** [comment]
@tdamsma I approve your work. Would you please review my docstring improvements and documentation additions in the [last commit](https://github.com/Pylons/deform/pull/486/commits/2ad4f952543490571d5a564f61ef759f3bdcdad2)? I think I got it right, but it helps to have a second pair of eyes. Thank you!

File: docs/widget.rst
Comment (tdamsma) [inline docs/widget.rst:334]: this seems like almost the same example as on line 303-309. Do we need both?

File: docs/widget.rst
Comment (tdamsma) [inline docs/widget.rst:260]:
```suggestion
When instantiating a new widget, you may specify its requirements by using the ``requirements`` attribute.
The requirements are specified as a sequence of two-tuples and/or dicts.
The two-tuple form is used by all core Deform widgets and uses the resource registry, taking advantage of its abstraction layer.
The dict form bypasses the resource registry.
The dict form may be easier to implement than the two-tuple form for custom widgets, as it does not introduce an implicit dependency on the registry. 
```

To be strictly correct the requirements is a sequence. The items in the sequence must be dicts or two-tuples.

Comment (stevepiercy) [reply docs/widget.rst:260]: Can they be a mixture of both types? I assumed that they are mutually exclusive, but after looking again at the code, it appears that they can be mixed.

File: deform/widget.py
Comment (tdamsma) [inline deform/widget.py:175]: this line is too long, not sure how to format it properly

**Review by tdamsma — COMMENTED** [review]
Nice improvements, please go ahead and finish it. I put some minor comments, feel free to ignore them

File: docs/widget.rst
Comment (stevepiercy) [reply docs/widget.rst:334]: It would be good to include a code example for both methods.

Nonetheless, this came as a surprise to me. I did not know that we could already specify `requirements` for a widget via a dict as shown in this example. Does this mean that the work you did is for naught?

File: deform/widget.py
Comment (stevepiercy) [reply deform/widget.py]: Good catch. Fixed in fd8860a.

File: docs/widget.rst
Comment (tdamsma) [reply docs/widget.rst:260]: yes, technically it can be mixed. Theoretically you an require some general version of jQuery from the registry and then some specific code for your widget

File: docs/widget.rst
Comment (tdamsma) [reply docs/widget.rst:334]: I added that example, there are now three examples (two tuples, one added by me, one by you). Think we only need two

Comment (stevepiercy) [reply docs/widget.rst]: My mistake. I thought that example pre-existed, but looking back at [commit history excluding my commits](https://github.com/Pylons/deform/pull/486/files/c77522fc4d1d8681379b1fb8288e7363a6d4839d), I see now. I split the narrative docs section into an introduction, and one subsection each for two-tuples and dicts, but didn't realize this.

Fixed in 88b4978 by combining the bits in pieces to flow better. @tdamsma Would you please review again?

File: docs/widget.rst
Comment (stevepiercy) [reply docs/widget.rst]: Got it. One more commit, and one more review from @tdamsma, and I think we can :shipit:!

See 0d4683a.

File: docs/widget.rst
Comment (tdamsma) [inline docs/widget.rst:259]: If you look carefully at this PR, you can spot that this is not true anymore. I realize now should have mentioned that explicitly in the PR, as you may very well disagree.

File: deform/widget.py
Comment (tdamsma) [inline deform/widget.py:1762]: I changed the requirements ere to the new form, as I feel this widget is tightly coupled to this bit of js. This is contradicting the statement 
> The two-tuple form is used by all core Deform widgets and uses the resource registry.

mentioned below

File: docs/widget.rst
Comment (tdamsma) [reply docs/widget.rst]: Great!

File: docs/widget.rst
Comment (tdamsma) [reply docs/widget.rst]: Much better!

File: docs/widget.rst
Comment (stevepiercy) [reply docs/widget.rst]: Good catch! Fixed in 8d25cc9.

File: deform/widget.py
Comment (stevepiercy) [reply deform/widget.py:1762]: Thanks! Fixed in 8d25cc9.

File: docs/widget.rst
Comment (stevepiercy) [reply docs/widget.rst]: Oh, and by the way, I agree. But then I realized that you changed the format for `fileupload`, but did not remove it from the default resource registry. Was that an oversight or deliberate?

File: deform/widget.py
Comment (tdamsma) [inline deform/widget.py:2206]:
```suggestion
```

File: docs/widget.rst
Comment (tdamsma) [reply docs/widget.rst]: sounds like an oversight

**Comment by stevepiercy** [comment]
I think we're done. w00t! Thank you @tdamsma!

**Pull request merged**
